R/nfunctions.R

```diff
@@ -176,15 +174,17 @@ ntype.corpus <- function(x, ...) {
 
 #' @export
 ntype.dfm <- function(x, ...) {
+    check_dots(list(...))
     x <- as.dfm(x)
-    ## only returns total non-zero features
+    # only returns total non-zero features
     result <- as.integer(rowSums(x > 0))
     names(result) <- docnames(x)
     result
 }
 
 #' @export
 ntype.tokens <- function(x, ...) {
+    if (length(list(...))) x <- tokens(x, ...)
     vapply(unclass(x), function(y) length(unique(y[y > 0])), integer(1))
```

Comment:
I think you should use this code for ntype.tokens()
```r
# nfeat.tokens <- function(x) {
#     if (attr(x, 'padding')) {
#         length(types(x)) + 1
#     } else {
#         length(types(x))
#     }
# }
```

Reply:
That would be fine except `ntype()` is vectorized over documents. The behaviour should be:
``` r
library("quanteda")
txt <- c(
  d1 = "a b c a b c",
  d2 = "a b c d e"
)
toks <- tokens(txt)

ntype(toks)
## d1 d2 
##  3  5
```
Also I don't think we should count the pad as a "type". So this is correct:
```r
ntype(tokens_remove(toks, "c", padding = TRUE))
## d1 d2 
##  2  4
```

Reply:
What function should people use to get the size of vocabulary in the entire corpus (tokens object)? Like `length(types(toks))`?

Reply:
That would be my recommendation.

So this thread is resolved and we can merge the PR?

Reply:
If unique types in tokens is equivalent to features in dfm, then `ntype() == nfeat()`. This is why I asked the questions.

Reply:
For clarity I’ll elaborate stuff we already know, but it answers the nfeat question. Here goes:

By our terminology and our implementation, tokens and types mean all words and unique words respectively, but these are attributes of documents. So, `ntoken()` and `ntype()` always return a named vector of the total tokens and total types, with each element corresponding to a document.

`tokens` objects have no “features”, only a dfm has features. That’s why we have not implemented `nfeat.tokens`. `nfeat()` only works on a dfm, and returns the number of columns. (Just as `ndoc.dfm()` returns the number of rows/documents.)

If someone wanted the unique types across a tokenised corpus, they could either use what you suggest in https://github.com/quanteda/quanteda/pull/1739#discussion_r307352760 or we could think of a new function as a shortcut. The existing nfunctions however are entirely consistent. Implementing a new one is not part of this PR.

Reply:
How about `ntoken(x, unique =TRUE)`?

Reply:
I think you mean `ntype(x, unique = TRUE)` but that's a not part of this issue and therefore not part of this PR, so we should be discussing and implementing it separately. 

Not good practice to hold up a specific fix because we see possible features to add when digging through the code for the specific fix.

Reply:
Ah, I though you are adding `ntype` in this PR that I even did not know existed.